iceoryx_dds/source/iceoryx_dds/dds/fastdds_data_reader.cpp

```diff
@@ -0,0 +1,287 @@
+    {
+        return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::NOT_CONNECTED);
+    }
+    if (buffer == nullptr)
+    {
+        return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::INVALID_RECV_BUFFER);
+    }
+
+    // get size of the sample
+    auto peekResult = peekNextSize();
+    if (peekResult.has_value())
+    {
+        uint64_t sampleSize = peekResult.value();
+        if (sampleSize == 0)
+        {
+            return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::INVALID_DATA);
+        }
+        if (bufferSize < sampleSize)
+        {
+            // Provided buffer is too small.
+            return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::RECV_BUFFER_TOO_SMALL);
+        }
+
+        // take maximum amount possible up to the cap: maxSamples
+        auto bufferCapacity = bufferSize / sampleSize;
+
+        auto numToTake = bufferCapacity;
+        if (maxSamples.has_value())
+        {
+            if (bufferCapacity > maxSamples.value())
+            {
+                numToTake = maxSamples.value();
+            }
+        }
+
+        uint64_t cursor = 0;
+        for (ulong i = 0; i < numToTake; i++)
+        {
+            if (!takeNext(&buffer[cursor], sampleSize))
+            {
+                break;
+            }
+            cursor += sampleSize;
+        }
+
+        uint64_t numSamplesBuffered = cursor / sampleSize;
+        return iox::cxx::success<uint64_t>(numSamplesBuffered);
+    }
+    else
+    {
+        return iox::cxx::success<uint64_t>(0u);
+    }
+}
+
+iox::capro::IdString_t iox::dds::FastDDSDataReader::getServiceId() const noexcept
+{
+    return m_serviceId;
+}
+
+iox::capro::IdString_t iox::dds::FastDDSDataReader::getInstanceId() const noexcept
+{
+    return m_instanceId;
+}
+
+iox::capro::IdString_t iox::dds::FastDDSDataReader::getEventId() const noexcept
+{
+    return m_eventId;
+}
+
+void iox::dds::FastDDSDataReader::waitForData(
+        eprosima::fastrtps::Duration_t maxWait)
+{
+    m_dataReader->wait_for_unread_message(maxWait);
+}
+
+void iox::dds::FastDDSDataReader::waitForWriterDiscovery(
+        uint16_t writersCount)
+{
+    std::unique_lock<std::mutex> lk(m_mutex);
+    m_writerDiscoveryCv.wait(
+        lk,
+        [&]
+        {
+            return (m_writerDiscoveryCount >= writersCount);
+        });
+
+    lk.unlock();
+    m_writerDiscoveryCv.notify_one();
+}
+
+
+iox::dds::FastDDSDataReader::FastDDSDataReaderListener::FastDDSDataReaderListener(
+        FastDDSDataReader* dataReader)
+    : m_fastddsDataReader(dataReader)
+{
+}
+
+iox::dds::FastDDSDataReader::FastDDSDataReaderListener::~FastDDSDataReaderListener()
+{
+}
+
+void iox::dds::FastDDSDataReader::FastDDSDataReaderListener::on_subscription_matched(
+        eprosima::fastdds::dds::DataReader*,
+        const eprosima::fastdds::dds::SubscriptionMatchedStatus& info)
+{
+    switch (info.current_count_change)
```

Comment:
If you like you could use also the ternary operator here and write:
```cpp
m_fastddsDataReader->m_writerDiscoveryCount += ( info.current_count_change == 1 ) ? +1 : -1;
m_fastddsDataReader->m_writerDiscoveryCv.notify.one();
```

Then you would have just two lines.